File: deob/src/main/java/dev/openrs2/deob/transform/BitShiftTransformer.java
package dev.openrs2.deob.transform;

import java.util.Set;

import dev.openrs2.asm.InsnMatcher;
import dev.openrs2.asm.InsnNodeUtils;
import dev.openrs2.asm.Transformer;
import dev.openrs2.asm.classpath.ClassPath;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.MethodNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class BitShiftTransformer extends Transformer {
	private static final Logger logger = LoggerFactory.getLogger(BitShiftTransformer.class);

	private static final InsnMatcher CONST_SHIFT_MATCHER = InsnMatcher.compile("(ICONST | BIPUSH | SIPUSH | LDC) (ISHL | ISHR | IUSHR | LSHL | LSHR | LUSHR)");
	private static final Set<Integer> LONG_SHIFTS = Set.of(Opcodes.LSHL, Opcodes.LSHR, Opcodes.LUSHR);

	private int simplified;

	@Override
	public void preTransform(ClassPath classPath) {
		simplified = 0;
	}

	@Override
	public void transformCode(ClassNode clazz, MethodNode method) {
		CONST_SHIFT_MATCHER.match(method).forEach(match -> {
			var push = match.get(0);
			var bits = InsnNodeUtils.getIntConstant(push);

			var opcode = match.get(1).getOpcode();
			var simplifiedBits = bits & (LONG_SHIFTS.contains(opcode) ? 63 : 31);

			if (bits != simplifiedBits) {
				method.instructions.set(push, InsnNodeUtils.createIntConstant(simplifiedBits));
				simplified++;
			}
		});
	}

	@Override
	public void postTransform(ClassPath classPath) {
		logger.info("Simplified {} bit shifts", simplified);
	}
}

File: deob/src/main/java/dev/openrs2/deob/transform/BitShiftTransformer.kt
package dev.openrs2.deob.transform

import com.github.michaelbull.logging.InlineLogger
import dev.openrs2.asm.InsnMatcher
import dev.openrs2.asm.classpath.ClassPath
import dev.openrs2.asm.classpath.Library
import dev.openrs2.asm.intConstant
import dev.openrs2.asm.toAbstractInsnNode
import dev.openrs2.asm.transform.Transformer
import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.MethodNode
import javax.inject.Singleton

@Singleton
class BitShiftTransformer : Transformer() {
    private var simplified = 0

    override fun preTransform(classPath: ClassPath) {
        simplified = 0
    }

    override fun transformCode(
        classPath: ClassPath,
        library: Library,
        clazz: ClassNode,
        method: MethodNode
    ): Boolean {
        for (match in CONST_SHIFT_MATCHER.match(method)) {
            val push = match[0]
            val bits = push.intConstant!!

            val opcode = match[1].opcode
            val mask = if (opcode in LONG_SHIFTS) 63 else 31

            val simplifiedBits = bits and mask

            if (bits != simplifiedBits) {
                method.instructions[push] = simplifiedBits.toAbstractInsnNode()
                simplified++
            }
        }
        return false
    }

    override fun postTransform(classPath: ClassPath) {
        logger.info { "Simplified $simplified bit shifts" }
    }

    companion object {
        private val logger = InlineLogger()
        private val CONST_SHIFT_MATCHER =
            InsnMatcher.compile("(ICONST | BIPUSH | SIPUSH | LDC) (ISHL | ISHR | IUSHR | LSHL | LSHR | LUSHR)")
        private val LONG_SHIFTS = setOf(Opcodes.LSHL, Opcodes.LSHR, Opcodes.LUSHR)
    }
}
